internal/oci/runtime_oci.go

```diff
@@ -1725,3 +1752,82 @@ func (r *runtimeOCI) checkpointRestoreSupported(runtimePath string) error {
 func (r *runtimeOCI) IsContainerAlive(c *Container) bool {
 	return c.Living() == nil
 }
+
+func (r *runtimeOCI) StartWatchContainerMonitor(ctx context.Context, c *Container) error {
+	if c.state.ContainerMonitorProcess != nil {
+		log.Debugf(ctx, "Skipping start watch container monitor for container %q: conmon process is nil. maybe the container have existed before cri-o update", c.ID())
+
+		return nil
+	}
+
+	// Check if the conmon process is the same process when the container was created.
+	conmonPid := c.state.ContainerMonitorProcess.Pid
+
+	startTime, err := getPidStartTime(conmonPid)
+	if err != nil {
+		return fmt.Errorf("failed to get conmon process start time: %w", err)
+	}
+
+	if c.state.ContainerMonitorProcess.StartTime != startTime {
+		return errors.New("conmon process has gone because the start time changed")
+	}
+
+	conmonProcess, err := os.FindProcess(conmonPid)
+	if err != nil {
+		return fmt.Errorf("couldn't find conmon process: %w", err)
+	}
+
+	go func() {
+		if c.watched {
+			// This shouldn't happen
+			log.Errorf(ctx, "Skipping start watch container monitor for container %q: conmon process is already watched", c.ID())
+		}
+
+		c.watched = true
+		defer func() {
+			c.watched = false
+		}()
+
+		// Wait until the conmon process exits
+		state, err := conmonProcess.Wait()
```

Comment:
we may want to do some perf scale testing to make sure this scales up with the number of containers. It may be more efficient to open the process and store it and periodically check it rather than have a goroutine open for each conmon

Reply:
I also think if we catch this situation happening we should emit a metric so we can track over time

Reply:
Thanks, don't we have a performance test CI or something, do we?

Reply:
no but we really should...

Reply:
I had hoped we could use pidfds for this behavior as non parent processes can open then and then the kernel will close them when the process dies, so we could asynchronously check whether the fd was closed (I believe this is how it works, it's possible I've got details fuzzy there)

Reply:
Thanks. I was trying to achieve it only with high-level libraries, but it should be possible with epoll or some way to mointor fds. I'll try implementing and compare the efficiency.